Classes/Fluid/ViewHelper/ComponentRenderer.php

```diff
@@ -279,6 +288,72 @@ public function prepareArguments()
         return $this->argumentDefinitions;
     }
 
+    /**
+     * Prepares a specified JSON schema for usage
+     *
+     * @param mixed $schema
+     * @return Schema
+     */
+    public function prepareSchema($schema)
+    {
+        $schemaFile = null;
+        if (is_array($schema) || is_object($schema)) {
+            // Normalize arrays and objects that come directly from fluid templates
+            $schema = json_encode($schema);
+            $hash = sha1($schema);
+        } else {
+            $schema = trim($schema);
+            if ($schema{0} === '{') {
+                // Seems to be JSON already
+                $hash = sha1($schema);
+            } else {
+                // JSON should be fetched from external file
+                $schemaFile = $this->resolveSchemaFile($schema);
+                $hash = sha1($schemaFile);
+            }
+        }
+
+        // Try to fetch it from cache first
+        if (isset(self::$argumentSchemaCache[$hash])) {
+            return self::$argumentSchemaCache[$hash];
+        }
+
+        // Fetch external schema file
+        if (isset($schemaFile)) {
+            if (!file_exists($schemaFile)) {
+                throw new SchemaException(
+                    sprintf(
+                        'The external schema definition "%s" specified in component "%s" could not be found: %s',
+                        $schema,
+                        $this->componentNamespace,
+                        $schemaFile
+                    ),
+                    1543314142
+                );
+            }
+            $schema = file_get_contents($schemaFile);
+        }
+
+        $schema = json_decode($schema);
+
+        // Check if JSON schema is valid JSON
+        if (json_last_error() !== JSON_ERROR_NONE) {
+            throw new SchemaException(
+                sprintf(
+                    'JSON schema definition produced "%s" in component "%s": %s.',
+                    json_last_error_msg(),
+                    $this->componentNamespace,
+                    $schemaFile ?? $schema
+                ),
+                1543322978
+            );
+        }
```

Comment:
I would extract line 323 to 349 to an seperate function called `retrieveSchemaFromFile(string $schemaFile)` to keep the code clean.